assayist/client/modify.py

```diff
@@ -0,0 +1,106 @@
+        if not isinstance(arg, str):
+            raise ValueError('c_name, c_type, and c_namespace must be strings')
+
+    if alternatives is None:
+        alternatives = []
+    if type(alternatives) not in (list, tuple, set):
+        raise ValueError('The alternatives keyword argument must be None or a list, tuple, or set')
+
+    for alternative in alternatives:
+        if not isinstance(alternative, str):
+            raise ValueError('The alternatives keyword argument must only contain strings')
+
+    if c_name in alternatives:
+        raise ValueError('The canonical name can\'t also be an alternative')
+
+    # Create a WHERE clause that checks to see if there is a component with the canonical name or
+    # alternative name with the passed in canonical name and alternatives
+    names_where_clause = [
+        'c.canonical_name = "{0}" OR "{0}" in c.alternative_names'.format(name)
+        for name in itertools.chain([c_name], alternatives)
+    ]
+    query = """
+    MATCH (c:Component {{canonical_type: "{}", canonical_namespace: "{}" }})
+    WHERE {}
+    RETURN c
+    """.format(c_type, c_namespace, ' OR '.join(names_where_clause))
+
+    results, _ = neomodel.db.cypher_query(query)
+    components = [Component.inflate(row[0]) for row in results]
+
+    # If no matching component is returned, just create one
+    if not components:
+        Component(canonical_namespace=c_namespace, canonical_name=c_name, canonical_type=c_type,
+                  alternative_names=alternatives).save()
+        return
+
+    for c in components:
+        # If this property is null in Neo4j, Neomodel doesn't treat it as an empty list, so just set
+        # it here to make the code below simpler since it can always assume it's a list
+        if c.alternative_names is None:
+            c.alternative_names = []
+
+    # This will be the only remaining component if there is more than one component returned, as
+    # the information stored in the others will be merged into this one
+    component = components[0]
+    if component.canonical_name != c_name:
+        # Add the old canonical name as an alternative name
+        if component.canonical_name not in component.alternative_names:
+            component.alternative_names.append(component.canonical_name)
+        # Remove the new canonical name from the alternative names if it's present
+        if c_name in component.alternative_names:
+            component.alternative_names.pop(component.alternative_names.index(c_name))
+        # Set the proper canonical name
+        component.canonical_name = c_name
+
+    # Add the missing alternative names
+    for alternative in alternatives:
+        if alternative not in component.alternative_names:
+            component.alternative_names.append(alternative)
+
+    # If there was only one component returned, then it's all set and no merging is required
+    if len(components) == 1:
+        component.save()
+        return
```

Comment:
This is not really necessary since if the components list is of size 1, the for-loop below will have zero iterations and will continue to saving the component.

Reply:
I had it there for readability, but I removed it.